tests/js/asserts.js

```diff
@@ -19,75 +19,122 @@
+            let message = `'${val1}' does not equal '${val2}' with tolerance '${tolerance}'`;
             if (errorMessage) {
-                message = errorMessage + ' - ' + message;
+                message = `${errorMessage} - ${message}`;
             }
-            throw new TestFailureError(message);
+            throw new TestFailureError(message, depth);
         }
     },
 
-    assertArray: function(value, length, errorMessage) {
+    assertArray: function(value, length, errorMessage, depth) {
         if (!Array.isArray(value)) {
-            throw new TestFailureError(errorMessage || `Value ${value} is not an array`);
+            throw new TestFailureError(errorMessage || `Value ${value} is not an array`, depth);
         }
     },
 
-    assertArrayLength: function(value, length, errorMessage) {
-        this.assertArray(value);
+    assertArrayLength: function(value, length, errorMessage, depth) {
+        this.assertArray(value, 1 + depth || 0);
         if (value.length !== length) {
-            throw new TestFailureError(errorMessage || `Value ${value} is not an array of length ${length}`);
+            throw new TestFailureError(errorMessage || `Value ${value} is not an array of length ${length}`, depth);
         }
     },
 
-    assertArraysEqual: function(val1, val2, errorMessage) {
-        var len1 = val1.length;
-        var len2 = val2.length;
-        var message;
+    assertArraysEqual: function(val1, val2, errorMessage, depth) {
+        this.assertDefined(val1, `val1 should be non-null but is ${val1}`, 1 + (depth || 0));
+        this.assertDefined(val2, `val2 should be non-null but is ${val2}`, 1 + (depth || 0));
+        const len1 = val1.length;
+        const len2 = val2.length;
 
         if (len1 !== len2) {
-            message = 'Arrays have different lengths (' + len1 + ' != ' + len2 + ')';
+            let message = `Arrays (${val1}) and (${val2}) have different lengths (${len1} != ${len2})`;
             if (errorMessage) {
-                message = errorMessage + ' - ' + message;
+                message = `${errorMessage} - ${message}`;
             }
-            throw new TestFailureError(message);
+            throw new TestFailureError(message, depth);
         }
 
-        for (var i = 0; i < len1; i++) {
-            if (val1[i] !== val2[i]) {
-                message = 'Array contents not equal at index ' + i + ' (' + val1[i] + ' != ' + val2[i] + ')';
+        let compare;
+        if (val1.type === "data") {
+            compare = (i, a, b) => a === b || this.assertArraysEqual(new Uint8Array(a), b, `Data elements at index ${i}`, 1) || true;
+        }
+        else if (val1.type === "date") {
+            compare = (i, a, b) => (a && a.getTime()) === (b && b.getTime());
+        }
+        else if (val1.type === "float" || val1.type === "double") {
+            compare = (i, a, b) => a >= b - 0.000001 && a <= b + 0.000001;
```

Comment:
If a and b are much larger than 1, this is equivalent to `a == b`. But it's fine as none of our tests use large value.